### PR TITLE
Implement `FromPyObject` and `IntoPy<PyObject>` traits for arrays (up to 32)

### LINKED ISSUE #775: `#[new]` should support arrays
Maybe a missing implementation?

```rust
use pyo3::prelude::*;

#[pyclass]
struct Foo {
  array: [i32; 2]
}

#[pymethods]
impl Foo {
  #[new]
  //pub fn new(array: [i32; 2]) -> Self {
  pub fn new(array: &[i32; 2]) -> Self {
    //Self { array }
    Self { array: *array }
  }
}
```

### PR DESCRIPTION
**Pull request by c410-f3r**
Fixes #775.

`Copy + Default` bounds to initialize memory for `copy_to_slice_impl` and avoid unsafe (`MaybeUninit` and `ptr` friends). `Copy` bound is also necessary due to the lack of a "real" `IntoIterator` implementation for arrays.

### CODE REVIEW

File: src/types/list.rs [hunk 1]
```diff
@@ -424,3 +464,10 @@ mod test {
         assert_eq!(2, list.get_item(3).extract::<i32>().unwrap());
     }
 }
+
+    }
+}
+
+    create_tests!(array);
+    create_tests!(vec vec!);
```
Comment (c410-f3r) [inline src/types/list.rs]: Not sure if this approach is desired. If so, I will also do the same in sequence.rs.

File: src/types/list.rs [hunk 2]
```diff
@@ -178,6 +178,42 @@ where
     }
 }
 
+macro_rules! array_impls {
+    ($($N:expr),+) => {
+        $(
+            impl<T> ToPyObject for [T; $N]
```
Comment (kngwyu) [inline src/types/list.rs]: As these imples are not faster than one for slices(`impl ~ for &[T]`), I don't think we need this.

File: src/types/sequence.rs [hunk 3]
```diff
@@ -282,6 +325,22 @@ where
     Ok(v)
 }
 
+fn extract_sequence_into_slice<'s, T>(obj: &'s PyAny, slice: &mut [T]) -> PyResult<()>
```
Comment (kngwyu) [inline src/types/sequence.rs]: How about `extract_sequence_inplace`?

**Comment by kngwyu** [comment]
Thank you!
`FromPyObject` impl looks good but I don't think we need `IntoPy<PyObject>` and `ToPyObject` impls for arrays.

**Comment by davidhewitt** [comment]
Also thanks from me!

> `FromPyObject` impl looks good but I don't think we need `IntoPy<PyObject>` and `ToPyObject` impls for arrays.

I slightly disagree with this. `ToPyObject` taking `&[T; N]` maybe doesn't offer benefit over `&[T]` but I believe `IntoPy<PyObject>` offers value.

For example, the following function will not compile without the `IntoPy<PyObject>` implementation:

    #[pyfunction]
    fn make_array()-> [i32; 2] {
        [0, 0]
    }

File: src/types/sequence.rs (hunk 3 above)
Comment (davidhewitt) [reply src/types/sequence.rs]: I tend to think `_inplace` as referring to the first argument, so I would vote against that name.

`extract_sequence_into_slice` is fine imo.

File: src/types/list.rs (hunk 2 above)
Comment (davidhewitt) [reply src/types/list.rs]: I agree for `ToPyObject` but not for `IntoPy<PyObject>` - see below.

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [0.9.0]
 
+* Implement `*Py` traits for arrays (up to 32).
```
Comment (davidhewitt) [inline CHANGELOG.md]: Can you please change the line to read the following and also move it to the `Added` section of the `0.9.0` changelog.

```suggestion
* Implement `FromPyObject` and `IntoPy<PyObject>` traits for arrays (up to 32). [#778](https://github.com/PyO3/pyo3/pull/778)
```

File: src/types/list.rs (hunk 1 above)
Comment (davidhewitt) [reply src/types/list.rs]: I think this is a good idea, thanks for adding comprehensive tests!

**Comment by c410-f3r** [comment]
Comments addressed. Thank you guys

**Comment by kngwyu** [comment]
> For example, the following function will not compile without the IntoPy<PyObject> implementation:

That makes sense, thanks :+1:

File: src/types/list.rs [hunk 5]
```diff
@@ -178,6 +178,33 @@ where
...
+        $(
+            impl<T> IntoPy<PyObject> for [T; $N]
+            where
+                T: Copy + IntoPy<PyObject>,
+            {
+                fn into_py(self, py: Python) -> PyObject {
+                    unsafe {
+                        let ptr = ffi::PyList_New(self.len() as Py_ssize_t);
+                        for (i, e) in self.iter().copied().enumerate() {
+                            let obj = e.into_py(py).into_ptr();
```
Comment (kngwyu) [inline src/types/list.rs]: `let obj = e.to_object(py)`.

File: src/types/sequence.rs [hunk 6]
```diff
@@ -282,6 +325,22 @@ where
...
+where
+    T: FromPyObject<'s>,
+{
+    let seq = <PySequence as PyTryFrom>::try_from(obj)?;
+    if (seq.len()? as usize) != slice.len() {
```
Comment (kngwyu) [inline src/types/sequence.rs]:
```suggestion
    if seq.len()? as usize != slice.len() {
```

File: src/types/sequence.rs [hunk 7]
```diff
@@ -330,6 +389,9 @@ impl<'v> PyTryFrom<'v> for PySequence {
 
 #[cfg(test)]
 mod test {
+    macro_rules! create_tests {
```
Comment (kngwyu) [inline src/types/sequence.rs]: Here not all tests are for `FromPyObject`.
Do we need to repeat all of them?
I think we really need it array version of `test_extract_bytearray_to_vec`.

File: src/types/list.rs [hunk 8]
```diff
@@ -205,6 +232,9 @@ where
 
 #[cfg(test)]
 mod test {
```
Comment (kngwyu) [inline src/types/list.rs]: Same as below.
Don't need to repeat all tests.

**Comment by kngwyu** [comment]
Thank you for addressing comments, but as I commented, I think current macro-generated tests seem to make it unclear what is tested here.
For `FromPyObject` implementation, we just need a test that calls `let arr: [u32; N] = obj.extract().unwrap()`. Most of tests like `test_seq_contains` are not related to this PR at all.
We can say the same thing for `IntoPy<PyObject>` implementation.
Most of the tests are not array-specific and use `ToPyObject for &[T]` implementation(e.g., `test_new`).
What we need is simply a test that calls `[1, 2, 3].into_py(py)`.

**Comment by c410-f3r** [comment]
I was waiting for David but looks like these tests are actually unwanted

**Comment by davidhewitt** [comment]
Thanks - sorry about the extra work there. I thought your macro test implementation was nice, but @kngwyu has been a maintainer of pyo3 for much longer than I so I'm still deferring to his opinion when we differ.

I wasn't going to ask you to undo the work though - was thinking about pushing a change myself. Thanks for doing so.

**Comment by kngwyu** [comment]
@c410-f3r 
Thank you for completing the PR!
@davidhewitt 
> so I'm still deferring to his opinion when we differ.

You don't have to do so. We sometimes make mistakes, but fortunately, we can help us with each other. When our opinion differs, it just means we need a discussion.

**Pull request merged**